Storage/SQL/SQLiteHistoryRecommendations.swift

```diff
@@ -82,8 +75,10 @@ extension SQLiteHistory: HistoryRecommendations {
             ]
         }
         
+        // The combined query clears the highlights cache and computes new highlights in one transaction
+        let combinedQuery = [clearHighlightsQuery, query].joined(separator: ";")
         // Run the highlights computation query and take the results to bulk insert into the cached highlights table
-        return self.db.runQuery(query, args: args, factory: argsFrom)
+        return self.db.runQuery(combinedQuery, args: args, factory: argsFrom)
             >>== { highlightRows in
```

Comment:
Ay, no. You're running a `SELECT` query, pulling all the values into memory, then stuffing them back into an array and inserting them into the cache table *in another transaction*. That second transaction (down on line 100) can fail. You've taken this code from three transactions down to two.

Use `INSERT INTO \(TableHighlights) SELECT …` to insert the values from the query directly into the table.

I think your argument for doing this is 'but we need to compute the cache key!'. But all of these values are derived from the URL, so you shouldn't need to pull everything into memory every time. Use a database function or persist them in the table or use a different cache key mechanism or calculate and memoize the cache key on display. All of these are better than taking two DB transactions and processing all of the rows to do this work.